refactor: migrate citations.py from pybtex to bibtexparser

gurukiran7 opened the pull request:
# Description

This PR refactors `pybamm/citations.py` to fully remove the legacy `pybtex` dependency and migrate all citation management to `bibtexparser` (v2). The new implementation:

- Uses bibtexparser for reading, parsing, and validating BibTeX entries
- Updates the `register` method to handle both known keys and raw BibTeX strings
- Adds improved error handling and type validation
- Implements custom citation string formatting to closely match the previous output
- Issues warnings when overwriting existing citations with different content


Fixes #3648



# Important checks:

Please confirm the following before marking the PR as ready for review:
- No style issues: `nox -s pre-commit`
- All tests pass: `nox -s tests`
- The documentation builds: `nox -s doctests`
- Code is commented for hard-to-understand areas
- Tests added that prove fix is effective or that feature works

agriyakhetarpal commented:
Thanks for starting this, @gurukiran7! To fix the tests, you'll need to add `bibtexparser` as in the `[cite]` set of optional dependencies in `pyproject.toml`. You may also run the tests locally to see if they pass and ease your development, as we'll need to trigger the workflows every time you push a commit: https://docs.pybamm.org/en/latest/source/user_guide/contributing.html#testing.

gurukiran7 commented:
@agriyakhetarpal I've updated the tests to align with the new citations.py implementation. At the moment, only a few tests are passing I'm still working on the rest. I want your feedback on the implementation so far

agriyakhetarpal commented:
Yes, that's alright, but the CI logs still don't reflect it, so it is difficult to see all the failures. I pulled your branch to my machine, and I see the same error: `ImportError: cannot import name 'BibDatabase' from 'bibtexparser'`. I think there might be a few commits on your local branch you haven't pushed?

https://github.com/search?q=repo:sciunto-org/python-bibtexparser+%22bibdatabase%22&type=code also reports no results, so I'm not sure where you're getting that API from.

It's not in the documentation either: https://bibtexparser.readthedocs.io/en/main/search.html?q=bibdatabase

agriyakhetarpal commented:
Yes, that's alright, but the CI logs still don't reflect it – it is hence difficult to see all the failures. I pulled your branch to my machine, and I see the same error: `ImportError: cannot import name 'BibDatabase' from 'bibtexparser'`. I think there might be a few commits on your local branch you haven't pushed?

https://github.com/search?q=repo:sciunto-org/python-bibtexparser+%22bibdatabase%22&type=code also reports no results, so I'm not sure where you're getting that attribute from.

It's not in the documentation either: https://bibtexparser.readthedocs.io/en/main/search.html?q=bibdatabase

gurukiran7 commented:
Thanks for pointing that out! the ImportError stems from my use of BibDatabase, which was part of bibtexparser v1 but has been removed in v2. I mistakenly mixed v1-style imports with the newer v2 API. I’ve now updated the code to use Entry from bibtexparser.model instead and removed all references to BibDatabase. That should fix the import issue  I’ll push the changes shortly.

gurukiran7 commented:
Hi @agriyakhetarpal , I've pushed updated code, and most methods and their tests are passing as expected However, there's an issue with the register method, causing tests like test_citations and test_overwrite_citation to fail. Despite debugging, I couldn't resolve  the issue, as it seems related to how citations are added to _all_citations and _papers_to_cite.
could u pls review the register method and suggest a fix?

agriyakhetarpal commented:
Thanks! Could you please run the whole test suite using `nox` directly? Running `pytest -k test_citations` is fine locally, but we're running all the tests in CI and that's different from running these tests in specific as you may see.

Also, you would need to fix the `style` job to allow the test suite to run.

kratman commented on src/pybamm/citations.py:
```diff
@@ -2,6 +2,10 @@
 import os
 import warnings
 from sys import _getframe
+import bibtexparser
+from bibtexparser import load as bibtex_load
+from bibtexparser import loads as bibtex_loads
+from bibtexparser.library import Library
```
I believe all the citation stuff was optional before, is there a reason it is not optional now?

gurukiran7 replied:
updated to optional

gurukiran7 commented:
@agriyakhetarpal I've pushed the latest updated code addressing most of the test failures and feedback

 Refactored the register method and _add_citation to support bibtexparser v2+
Made citation handling optional using import_optional_dependency
 Cleaned up all formatting/style issues (ruff-format, pre-commit now passes)
  if you have any suggestions on expected behavior or test intentn pls let me know

gurukiran7 commented:
@agriyakhetarpal , @kratman , I've pushed the latest code please review when you get the chance
